flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushFilterInCalcIntoTableSourceScanRule.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.planner.plan.rules.logical;
+
+import org.apache.flink.table.api.TableConfig;
+import org.apache.flink.table.api.config.OptimizerConfigOptions;
+import org.apache.flink.table.connector.source.abilities.SupportsFilterPushDown;
+import org.apache.flink.table.planner.plan.nodes.logical.FlinkLogicalCalc;
+import org.apache.flink.table.planner.plan.nodes.logical.FlinkLogicalTableSourceScan;
+import org.apache.flink.table.planner.plan.schema.FlinkPreparingTableBase;
+import org.apache.flink.table.planner.plan.schema.TableSourceTable;
+import org.apache.flink.table.planner.utils.ShortcutUtils;
+
+import org.apache.calcite.plan.RelOptRuleCall;
+import org.apache.calcite.rel.logical.LogicalTableScan;
+import org.apache.calcite.rex.RexNode;
+import org.apache.calcite.rex.RexProgram;
+import org.apache.calcite.rex.RexProgramBuilder;
+import org.apache.calcite.tools.RelBuilder;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import scala.Tuple2;
+
+/**
+ * Pushes a {@link org.apache.calcite.rel.logical.LogicalFilter} from the {@link org.apache.calcite.rel.logical.LogicalCalc}
+ * and into a {@link LogicalTableScan}
+ */
+public class PushFilterInCalcIntoTableSourceScanRule extends PushFilterIntoSourceScanRuleBase {
+    public static final PushFilterInCalcIntoTableSourceScanRule INSTANCE =
+            new PushFilterInCalcIntoTableSourceScanRule();
+
+    public PushFilterInCalcIntoTableSourceScanRule() {
+        super(
+                operand(FlinkLogicalCalc.class, operand(FlinkLogicalTableSourceScan.class, none())),
+                "PushFilterInCalcIntoTableSourceScanRule");
+    }
+
+    @Override
+    public boolean matches(RelOptRuleCall call) {
+        TableConfig config = ShortcutUtils.unwrapContext(call.getPlanner()).getTableConfig();
+        if (!config.getConfiguration()
+                .getBoolean(
+                        OptimizerConfigOptions.TABLE_OPTIMIZER_SOURCE_PREDICATE_PUSHDOWN_ENABLED)) {
+            return false;
+        }
```

Comment:
Add these codes to `canPushdownFilter`?

Reply:
This is used as an early return for both this class and the filter for pushing into the source. The problem is that the logic below which actually extracts the source table is different for both implementations, hence why I didn't move this to `canPushdownFilter`.

Reply:
What about moving to `PushFilterIntoSourceScanRuleBase#matches`? We can use `super.matches` here to check.